crypto/openssl_csr.py

```diff
@@ -0,0 +1,299 @@
+subjectAltName:
+    description: The alternative names this CSR is valid for
+    returned:
+        - changed
+        - success
+    type: string
+    sample: 'DNS:www.ansible.com,DNS:m.ansible.com'
+'''
+
+class CertificateSigningRequestError(Exception):
+    pass
+
+class CertificateSigningRequest(object):
+
+    def __init__(self, module):
+        self.state = module.params['state']
+        self.digest = module.params['digest']
+        self.force = module.params['force']
+        self.subjectAltName = module.params['subjectAltName']
+        self.path = module.params['path']
+        self.privatekey_path = module.params['privatekey_path']
+        self.version = module.params['version']
+        self.changed = True
+        self.request = None
+        self.privatekey = None
+
+        self.subject = {
+          'C': module.params['countryName'],
+          'ST': module.params['stateOrProvinceName'],
+          'L': module.params['localityName'],
+          'O': module.params['organizationName'],
+          'OU': module.params['organizationalUnitName'],
+          'CN': module.params['commonName'],
+          'emailAddress': module.params['emailAddress'],
+        }
+
+        if self.subjectAltName is None:
+            self.subjectAltName = 'DNS:%s' % self.subject['CN']
+
+        for (key,value) in self.subject.items():
+            if value is None:
+                del self.subject[key]
+         
+
+    @property
+    def file_path(self):
+        return self.path
+   
+
+    def generate(self, module):
+        '''Generate the certificate signing request.'''
+
+        if not os.path.exists(self.file_path) or self.force:
+            req = crypto.X509Req()
+            req.set_version(self.version)
+            subject = req.get_subject()
+            for (key,value) in self.subject.items():
+                if value is not None:
+                    setattr(subject, key, value)
+
+            if self.subjectAltName is not None:
+                req.add_extensions([crypto.X509Extension("subjectAltName", False, self.subjectAltName)])
+
+            privatekey_content = open(self.privatekey_path).read()
+            self.privatekey = crypto.load_privatekey(crypto.FILETYPE_PEM, privatekey_content)
+
+            req.set_pubkey(self.privatekey)
+            req.sign(self.privatekey, self.digest)
+            self.request = req
+
+            csr_file = open(self.file_path, 'w')
+            csr_file.write(crypto.dump_certificate_request(crypto.FILETYPE_PEM, self.request))
+            csr_file.close()
+        else:
+            self.changed = False
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+        
+
+    def remove(self):
+        '''Remove the Certificate Signing Request.'''
+
+        try:
+            os.remove(self.file_path)
+        except OSError:
```

Comment:
This would ignore if the file can't be removed, see https://github.com/ansible/ansible-modules-extras/blob/devel/crypto/openssl_privatekey.py#L168